File: reveal-sc2-opponent-main/stream-overlay-backend/lib/data_integrity_metrics.js
// @ts-check
/**
 * lib/data_integrity_metrics.js -- Stage 7 of STAGE_DATA_INTEGRITY_ROADMAP.
 *
 * Process-local counters + duration histograms for the JS-side
 * write hot paths (analyzer.js persistMetaDb, the routes/* atomic
 * writers). Mirrors core/data_integrity_metrics.py byte-for-byte
 * on counter names so the SPA can render a single dashboard widget
 * regardless of which writer fired the event.
 *
 * Why not Prometheus / OpenTelemetry: sc2tools is a single-user
 * desktop app; a 50MB scraper for one process is wrong-shaped.
 * The Stage 14 master-roadmap Sentry hook can later subscribe
 * to .error() events without changing this surface.
 */

'use strict';

const COUNTER_NAMES = Object.freeze([
  'write_attempted',
  'write_succeeded',
  'write_failed',
  'lock_acquired',
  'lock_contended',
  'lock_timeout',
  'validation_rejected',
  'salvage_triggered',
  'recovery_staged',
  'recovery_applied',
  'schema_too_new_rejection',
]);

const _state = {
  counters: {},      // counter -> basename -> int
  histograms: {},    // (name + ':' + basename) -> number[]
  errors: [],
  maxErrors: 100,
  histogramCap: 1000,
};

function _ensureCounter(name, basename) {
  if (!_state.counters[name]) _state.counters[name] = {};
  if (!_state.counters[name][basename]) _state.counters[name][basename] = 0;
}

/**
 * Increment a named counter. Unknown names are ignored after a
 * console warning (the lint rule below surfaces typos at PR time).
 *
 * @param {string} name
 * @param {{ basename?: string, n?: number }} [opts]
 */
function counterInc(name, opts) {
  if (!COUNTER_NAMES.includes(name)) {
    console.warn(`[metrics] unknown counter: ${name}`);
    return;
  }
  const basename = (opts && opts.basename) || '_global';
  const n = (opts && typeof opts.n === 'number') ? opts.n : 1;
  _ensureCounter(name, basename);
  _state.counters[name][basename] += n;
}

/**
 * Record a duration in ms. Values past the histogramCap are evicted
 * FIFO so a long-running process never grows unbounded memory.
 *
 * @param {string} name
 * @param {number} value
 * @param {{ basename?: string }} [opts]
 */
function histogramRecord(name, value, opts) {
  if (typeof value !== 'number' || !isFinite(value)) return;
  const basename = (opts && opts.basename) || '_global';
  const key = name + ':' + basename;
  if (!_state.histograms[key]) _state.histograms[key] = [];
  _state.histograms[key].push(value);
  if (_state.histograms[key].length > _state.histogramCap) {
    _state.histograms[key] = _state.histograms[key]
      .slice(-_state.histogramCap);
  }
}

/**
 * Record a structured error. Pinned in a ring buffer so the
 * diagnostics widget can render the most-recent N events.
 *
 * @param {string} kind
 * @param {{ detail?: object }} [opts]
 */
function error(kind, opts) {
  const record = {
    kind,
    detail: (opts && opts.detail) || {},
    timestamp: Date.now() / 1000,
  };
  _state.errors.push(record);
  if (_state.errors.length > _state.maxErrors) {
    _state.errors = _state.errors.slice(-_state.maxErrors);
  }
  console.error('[data_integrity error]', kind, record.detail);
}

function _summarize(vals) {
  if (!vals.length) return { count: 0 };
  const sorted = vals.slice().sort((a, b) => a - b);
  const n = sorted.length;
  const pct = (p) => sorted[Math.max(0, Math.min(n - 1, Math.floor(p * (n - 1))))];
  return {
    count: n,
    p50: pct(0.50),
    p95: pct(0.95),
    p99: pct(0.99),
    min: sorted[0],
    max: sorted[n - 1],
  };
}

/**
 * @returns {{
 *   counters: object,
 *   histogram_summary: object,
 *   recent_errors: object[]
 * }}
 */
function snapshot() {
  const histSummary = {};
  for (const k of Object.keys(_state.histograms)) {
    histSummary[k] = _summarize(_state.histograms[k]);
  }
  return {
    counters: JSON.parse(JSON.stringify(_state.counters)),
    histogram_summary: histSummary,
    recent_errors: _state.errors.slice(),
  };
}

function _resetForTests() {
  _state.counters = {};
  _state.histograms = {};
  _state.errors = [];
}

module.exports = {
  COUNTER_NAMES,
  counterInc,
  histogramRecord,
  error,
  snapshot,
  _resetForTests,
};

File: reveal-sc2-opponent-main/core/data_integrity_metrics.py
"""
core.data_integrity_metrics -- Stage 7 of STAGE_DATA_INTEGRITY_ROADMAP.

Lightweight in-process counters + structured logging for the
data-write hot paths. Counters are read by ``routes/diagnostics.js``
to render the "write health" widget on the SPA's Diagnostics tab.

Why an in-process module rather than ``prometheus_client``
----------------------------------------------------------
sc2tools runs as a single Node process plus a single Python watcher
on the user's desktop -- there is no Prometheus scraper, no shared
metrics endpoint, no Grafana. We need cheap counters that any caller
can increment and the diagnostics page can read in one HTTP round
trip. Layering Prometheus on top would mean a 50 MB extra dependency
for a single-user app.

Sentry hook
-----------
Stage 14 of the master roadmap ships Sentry. Until then the
``error()`` function is a structured ``logger.error`` call so the
existing log viewer / debug bundle picks up the same fields.

Usage
-----

    from core.data_integrity_metrics import DataIntegrityMetrics
    metrics = DataIntegrityMetrics.singleton()
    metrics.counter_inc("write_attempted", basename="MyOpponentHistory.json")
    metrics.histogram_record("write_duration_ms",
                             123, basename="MyOpponentHistory.json")
    metrics.error("DataIntegrityError",
                  detail={"basename": "...", "violation": "..."})

The Express backend reads via the JS sibling
``stream-overlay-backend/lib/data_integrity_metrics.js`` which is
process-local. The diagnostics widget calls
``GET /api/data-integrity/metrics`` to surface them.
"""

from __future__ import annotations

import logging
import threading
import time
from collections import Counter, defaultdict
from typing import Any, Dict, List, Optional

logger = logging.getLogger("data_integrity")


COUNTER_NAMES = (
    "write_attempted",
    "write_succeeded",
    "write_failed",
    "lock_acquired",
    "lock_contended",
    "lock_timeout",
    "validation_rejected",
    "salvage_triggered",
    "recovery_staged",
    "recovery_applied",
    "schema_too_new_rejection",
)


class DataIntegrityMetrics:
    """Process-local counters + duration histograms.

    Thread-safe: a single RLock guards every mutation. Reads return
    snapshots so callers don't see partial state.
    """

    _instance: Optional["DataIntegrityMetrics"] = None
    _instance_lock = threading.Lock()

    def __init__(self) -> None:
        self._lock = threading.RLock()
        self._counters: Dict[str, Counter] = defaultdict(Counter)
        self._histograms: Dict[str, List[float]] = defaultdict(list)
        self._errors: List[Dict[str, Any]] = []
        self._max_errors = 100  # ring buffer for the diagnostics widget

    @classmethod
    def singleton(cls) -> "DataIntegrityMetrics":
        with cls._instance_lock:
            if cls._instance is None:
                cls._instance = cls()
            return cls._instance

    def counter_inc(self, name: str, *, basename: str = "_global", n: int = 1) -> None:
        if name not in COUNTER_NAMES:
            logger.warning("metrics: unknown counter %s", name)
            return
        with self._lock:
            self._counters[name][basename] += n

    def histogram_record(
        self, name: str, value: float, *, basename: str = "_global"
    ) -> None:
        if not isinstance(value, (int, float)):
            return
        with self._lock:
            self._histograms[name + ":" + basename].append(float(value))
            # Keep the histogram bounded so a long-running process
            # doesn't accumulate unbounded memory.
            if len(self._histograms[name + ":" + basename]) > 1000:
                self._histograms[name + ":" + basename] = (
                    self._histograms[name + ":" + basename][-1000:]
                )

    def error(self, kind: str, *, detail: Optional[Dict[str, Any]] = None) -> None:
        record = {
            "kind": kind,
            "detail": detail or {},
            "timestamp": time.time(),
        }
        with self._lock:
            self._errors.append(record)
            if len(self._errors) > self._max_errors:
                self._errors = self._errors[-self._max_errors:]
        logger.error(
            "data_integrity error", extra={"event": "data_integrity_error", **record}
        )

    def snapshot(self) -> Dict[str, Any]:
        with self._lock:
            return {
                "counters": {
                    name: dict(self._counters[name])
                    for name in COUNTER_NAMES
                    if name in self._counters
                },
                "histogram_summary": {
                    key: _summarize(vals)
                    for key, vals in self._histograms.items()
                },
                "recent_errors": list(self._errors),
            }


def _summarize(vals: List[float]) -> Dict[str, float]:
    if not vals:
        return {"count": 0}
    sorted_vals = sorted(vals)
    n = len(sorted_vals)
    def pct(p: float) -> float:
        idx = max(0, min(n - 1, int(p * (n - 1))))
        return sorted_vals[idx]
    return {
        "count": n,
        "p50": pct(0.50),
        "p95": pct(0.95),
        "p99": pct(0.99),
        "min": sorted_vals[0],
        "max": sorted_vals[-1],
    }


__all__ = ["DataIntegrityMetrics", "COUNTER_NAMES"]
